doc/source/tutorial/optimize.rst

```diff
@@ -440,89 +440,225 @@ example using the Rosenbrock function follows:
 
-    \begin{eqnarray*}
-      x^3 - y &= 0 \\
-      y - 1 &\geq 0
-    \end{eqnarray*}
+     \begin{equation*} \left[\begin{array}{c}-\infty \\1\end{array}\right] \leq
+      \left[\begin{array}{cc} 1& 2 \\ 2& 1\end{array}\right]
+       \left[\begin{array}{c} x_0 \\x_1\end{array}\right] \leq
+        \left[\begin{array}{c} 1 \\ 1\end{array}\right],\end{equation*}
 
+and defined using:
 
+    >>> from scipy.optimize import LinearConstraint
+    >>> lin_cons = LinearConstraint([[1, 2], [2, 1]], [-np.inf, 1], [1, 1])
 
-The objective function and its derivative are defined as follows.
+Defining Nonlinear Constraints:
+"""""""""""""""""""""""""""""""
+The nonlinear constraints are defined using :func:`NonlinearConstraint` object.
+For the example, the nonlinear constraints :math:`x_0^2 + x_1 \leq 1`
+and :math:`x_0^2 - x_1 \leq 1` may be defined as:
 
-    >>> def func(x, sign=1.0):
-    ...     """ Objective function """
-    ...     return sign*(2*x[0]*x[1] + 2*x[0] - x[0]**2 - 2*x[1]**2)
+    >>> from scipy.optimize import NonlinearConstraint
+    >>> def cons_f(x):
+    ...     return [x[0]**2 + x[1], x[0]**2 - x[1]]
+    >>> def cons_J(x):
+    ...     return [[2*x[0], 1], [2*x[0], -1]]
+    >>> def cons_H(x, v):
+    ...     return v[0]*np.array([[2, 0], [0, 0]]) + v[1]*np.array([[2, 0], [0, 0]])
+    >>> nonl_cons = NonlinearConstraint(cons_f, -np.inf, 1, jac=cons_J, hess=cons_H)
 
-    >>> def func_deriv(x, sign=1.0):
-    ...     """ Derivative of objective function """
-    ...     dfdx0 = sign*(-2*x[0] + 2*x[1] + 2)
-    ...     dfdx1 = sign*(2*x[0] - 4*x[1])
-    ...     return np.array([ dfdx0, dfdx1 ])
+Computing the second derivatives of the constraints may be too troublesome to implement
+or computationally too expensive. In this case, the second derivatives may be approximated
+using a :func:`HessianUpdateStrategy`
+object, which implements quasi-Newton approximations. For instance,  :func:`BFGS`
+could be used to approximate the Hessian:
 
-Note that since :func:`minimize` only minimizes functions, the ``sign``
-parameter is introduced to multiply the objective function (and its
-derivative) by -1 in order to perform a maximization.
+    >>> from scipy.optimize import BFGS
+    >>> nonl_cons = NonlinearConstraint(cons_f, -np.inf, 1, jac=cons_J, hess=BFGS())
 
-Then constraints are defined as a sequence of dictionaries, with keys
-``type``, ``fun`` and ``jac``.
+Alternatively, the Hessian may also be approximated using finite-differences:
 
-    >>> cons = ({'type': 'eq',
-    ...          'fun' : lambda x: np.array([x[0]**3 - x[1]]),
-    ...          'jac' : lambda x: np.array([3.0*(x[0]**2.0), -1.0])},
-    ...         {'type': 'ineq',
-    ...          'fun' : lambda x: np.array([x[1] - 1]),
-    ...          'jac' : lambda x: np.array([0.0, 1.0])})
+    >>> nonl_cons = NonlinearConstraint(cons_f, -np.inf, 1, jac=cons_J, hess='2-point')
 
+If the constraint first derivatives are also not known. They may be approximated
```

Comment:
`The Jacobian of the constraints can be approximated by finite differences as well:`

Reply:
Ok